src/Microsoft.OData.Core/UriParser/Binders/SelectExpandBinder.cs

```diff
@@ -236,85 +268,340 @@ private SelectItem GenerateExpandItem(ExpandTermToken tokenIn)
             }
 
             // Add the segments in select and expand to parsed segments
-            this.parsedSegments.AddRange(pathSoFar);
+            List<ODataPathSegment> parsedPath = new List<ODataPathSegment>(this.parsedSegments);
+            parsedPath = parsedPath.Concat(pathSoFar).ToList();

```

Comment:
Cheaper: 
    List<ODataPathSegment> parsedPath = new List<ODataPathSegment>(this.parsedSegments.Concat(pathSoFar));